### PR TITLE
Fix: TF Agent host charm rebuilds COS scrape jobs based on supervisor config.

### PR DESCRIPTION
**Pull request by gntzio**
## Description

After deploying the Grafana Agent hosts in TEL, the `/etc/grafana-agent.yaml` config was not properly populated with scrape endpoints.

Following investigation, the suspicion is that since `charm.py` is instantiated on every Juju event, the in-memory `self.scrape_endpoints` variable is not consistently populated.

To address this, `COSAgentProvider.scrape_configs` are now sourced directly from the `supervisord` configuration.

## Resolved issues

[CERTTF-1038](https://warthogs.atlassian.net/browse/CERTTF-1038)

## Documentation

Charm unit tests.

## Web service API changes

N/A

## Tests

Charm unit tests.

[CERTTF-1038]: https://warthogs.atlassian.net/browse/CERTTF-1038?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py
Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:37]:
```suggestion
        for conf_file in [c for supervisor_conf_dir.iterdir() if c.suffix == ".conf"]:
```
Consider moving the if block up, I think it make is more obvious that the loop is only considering .config files.  Alternatively use glob.

Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:38]:
```suggestion
```
See above, would also change indentation on lines 39-69

**Review by rene-oromtz — APPROVED** [review]
LGTM, as far as I remember, Grafana Agent required the relation with Prometheus to actually configure the scrape jobs. 

Due to initial TEL deployment without Prometheus, that relation data may never got populated from our charm side... I think the key is actually the `refresh_events` addition which also makes sense because when we add new agents, we should also be refreshing the scrape_jobs with the new agents config.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py
Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:42]: What if the config doesn't have an agent name and port -- it seems that we will just stkoip them  Is that OK or should that be an error condition?

Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:49]: The OSError and IOError tracebacks that mght have been produced by parse_supervisor_config_file are caught within parse_supervisor_config_file.

Please consider adding tests for the failure conditions and new code.

Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:54]: The ValueError that parse_supervisor_config_file is already handled within parse_supervisor_config_file and causes it to return None, None

Same for the

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py [hunk 1]
```diff
@@ -64,6 +64,24 @@ def get_supervisor_agents_port_mapping() -> dict[str, int]:
...
+    """
+    return [
+        {
+            "job_name": agent_name,
+            "metrics_path": "/metrics",
+            "static_configs": [{"targets": [f"localhost:{port}"]}],
+        }
+        for agent_name, port in sorted(
+            get_supervisor_agents_port_mapping().items()
+        )
```
Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]:
```suggestion
        ) if agent_name
```

Do we want to include "None" in this list?

Need to test behavior if the value were to come back None anyway

**Review by ajzobro — CHANGES_REQUESTED** [review]

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py
Comment (rene-oromtz) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:37]: As most of these comments are not related to the code change, do you think we should open a Jira card to track this improvements you already identified? This may be better suited to address in a followup PR if they are not directly impacting the code changes.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (gntzio) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: I would leave the logic "as is" to avoid semantic complexity. 
`get_supervisor_agents_port_mapping` should not by design return `(None, None)`.
In other case, other consumers should perform the same check (e.g.in `charm.py`).

**Comment by gntzio** [comment]
@ajzobro, thanks for pointing out potential improvements.
As far as I can see, they are (mostly) not related to the changes in this PR itself. 
I would appreciate it, if these could be addressed in a separate "improvements" PR.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (rene-oromtz) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: Can it be None? It defaults to empty dict {} and only inserts to the agent_port_mapping if agent and port was found.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py
Comment (rene-oromtz) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:37]: I went ahead and open [CERTTF-1041](https://warthogs.atlassian.net/browse/CERTTF-1041) please let me know what you think!

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (ajzobro) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: If "by design" you mean that every agent will always have a config file, that config file can be read and the values within the config file for agent name and port will always be defined: then why do we have any error handling in parse_supervisor_config_file at all?   Most of the return statetments return None, None -- if we're going to make the argument that by design we never end up in any of those failure paths, then why do they exist?

This new code that you wrote is definitely in scope, and does not handle all possible returns from parse_supervisor_config_file in a way that doesn't just pass along the problem.

**Review by ajzobro — CHANGES_REQUESTED** [review]
Please address the issues that can easily be addressed within the scope of the code that was introduced here, including tests for the failure conditions.

CERTTF-1041 now eixsts as a strategy for handling (kicking along) the technical debt this time.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (gntzio) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: `parse_supervisor_config_file()` can return `(None, None)`, but `get_supervisor_scrape_jobs()` does not consume that output directly. `get_supervisor_agents_port_mapping()` already filters invalid parser results and returns only valid agent -> port entries, as stated in its docstring.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py
Comment (ajzobro) [inline agent/charms/testflinger-agent-host-charm/src/supervisord.py:48]: we can probably break after this unless we want to continue processing files and collect all the egent configs and take the last defined port for any config file for this agent.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (ajzobro) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: > Can it be None? It defaults to empty dict {} and only inserts to the agent_port_mapping if agent and port was found.

Indeed, `get_supervisor_agents_port_mapping` cleans that up and returns an empty dict; so this will just crash.

>>> [{'a':'b','c':x,'d':y} for x, y in sorted( [ {} ])]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: not enough values to unpack (expected 2, got 0)
>>>

**Review by ajzobro — COMMENTED** [review]
If we write the tests for the error cases we will clearly see the issues in the code.

File: agent/charms/testflinger-agent-host-charm/src/supervisord.py (hunk 1 above)
Comment (gntzio) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: The example is not equivalent to this code path: we iterate over `sorted(get_supervisor_agents_port_mapping().items())`. 
So the equivalent expression would be `[{'a': 'b', 'c': x, 'd': y} for x, y in sorted({}.items())]`, which evaluates to `[]`.

Comment (rene-oromtz) [reply agent/charms/testflinger-agent-host-charm/src/supervisord.py:81]: That is failing because of this `[ {} ])` . you should be using: `[{'a':'b','c':x,'d':y} for x, y in sorted([])]` or you can mock this directly:
```
>>> agent_port_mapping = {}
>>> scrape_config = [
...         {
...             "job_name": agent_name,
...             "metrics_path": "/metrics",
...             "static_configs": [{"targets": [f"localhost:{port}"]}],
...         }
...         for agent_name, port in sorted(
...             agent_port_mapping.items()
...         )
... ]
>>> 
>>> print(scrape_config)
[]
```
Ah Gena was faster in replying

**Comment by ajzobro** [comment]
I guess I'm not getting any more tests, huh?

**Review by ajzobro — APPROVED** [review]
Agree with creation of 1041 for improved testing and fixing the relied-upon-but-out-of-scope functions.

**Comment by rene-oromtz** [comment]
Thanks @ajzobro! Agree that there is lots to improve. At least now you already identify what needs to be done it should be a bit faster to implement in 1041
